### ISSUE TITLE
[Enhancement] Bottom Sheet/drawer feature request

### ISSUE THREAD

**Issue by developer9969** [xamarin/XamarinCommunityToolkit#362]
## Summary
Hi all,
Fantastic that now we have this toolkit!!!
Many applications today have ability to show Bottom Sheet/drawer  see pic below
![image](https://user-images.githubusercontent.com/16970109/94985430-94fb0980-054e-11eb-9591-8ba199452d84.png)

David Ortinau  created a small sample here 
https://github.com/davidortinau/SampleBottomDrawer

![image](https://user-images.githubusercontent.com/16970109/94985444-c1168a80-054e-11eb-8777-aa41212cc173.png)

It would be fantastic if we could have a control out of the box that works in both iOS/Android and is draggable up and down.

Would you consider including such a control?

many thanks

**Comment by zack7891** [comment]
This would be great to add to the toolkit I might just take this on.

**Comment by Hackavist** [comment]
I think it would add great value i would like to contribute to it as well

**Comment by GiampaoloGabba** [comment]
I think we need a spec.
For example, a good bottom-sheet control provides:
- multiple steps
- constraints
- multiple gesture recognizer to work when there is a scrollview inside
- overlay management (background color, click/swipe to close on overlay...)
- ecc...

Also we need to choose if we are using pure forms view or natives (for example on ios putting a UIViewcontroller inside the popup to have nested navigation.

I have already done something similar (without multiple steps) but is based on rg.popups so i think is not a viable option for this Toolkit :(

I can help with the spec, then we can implement it :)

**Comment by GiampaoloGabba** [comment]
Here some (iOS) examples of what i'm talking about:

https://github.com/OfTheWolf/UBottomSheet
https://github.com/gordontucker/FittedSheets
https://github.com/hryk224/Bottomsheet

**Comment by GiampaoloGabba** [comment]
Last thing: there are quite a few implementations of this in github. all of them are using RelativeLayout.
I'm against it: it's slow as hell... soooo please dont use it :)

**Comment by Hackavist** [comment]
i can see what you are trying to do here, I think we can have a bindable property of a view so you can add whatever view you want inside but the idea (what i can't image) is how can we make it resizable like [this](https://github.com/hryk224/Bottomsheet) i get the idea but i can't think of a way to implement it

**Comment by GiampaoloGabba** [comment]
the difficult thing is too coordinate the steps with the content inside.
Let's see what the maintainers think about this, i'm sure they will come with some super-neat ideas :)

**Comment by Hackavist** [comment]
Sounds Great i am excited

**Comment by KaranRami** [comment]
I have a similar requirement in my project, and I achieved using Rg popup. It is purely done with what I had available. One can find my sample project on GitHub: https://github.com/KaranRami/CustomAlerts

So meanwhile someone comes with out of the box solution, people can try to fit in with their requirement.

**Comment by Tommigun1980** [comment]
These are some components I made a while back that accomplish this:
https://github.com/Tommigun1980/Xamarin.Forms.DragView
https://github.com/Tommigun1980/Xamarin.Forms.SlideView

**Comment by Hackavist** [comment]
@Tommigun1980 Well done these are really inspiring and good looking all we need now is the maintainers approval

**Comment by Hackavist** [comment]
@jfversluis can we start the process implementing this issue?

**Comment by Tommigun1980** [comment]
> @Tommigun1980 Well done these are really inspiring and good looking all we need now is the maintainers approval

Thanks. The DragView judders on Android though, I would have to fix it first in case there's interest in these.

**Comment by jfversluis** [comment]
@Hackavist I think the idea is great, but I'd like to think a little bit more about the implementation of this before we start diving into this. As far as I am concerned you could start doing this, but don't put too much effort in it just yet.

I'd like to see some kind of a spec on how you would see this working first :) Does that makes sense?

@Tommigun1980 just to be clear, are you OK if people take (pieces of) your code as a base? Also for the other plugins for this mentioned, if you are going to use any of that, please check the license and contact the author/maintainer for consent.

**Comment by ahoefling** [comment]
Thank you @developer9969 for suggesting this idea. I think it will be a great addition to the platform. Once we merge #292 the implementation of this is trivial. I have built a very similar control using the popup control that is scheduled to be PRed once we support 5.0.0 of Xamarin.Forms

**Comment by roubachof** [comment]
@ahoefling I don’t think a native pop up and a bottom sheet have the same semantic. 

For example one would expect several states like:
* hidden state
* collapsed
* expanded 

Also it should react to swipe up and down to navigate through state, as well as drag gesture.

A good reference of what to expect from a bottom sheet is here: 
https://material.io/components/sheets-bottom/#bottom-sheets-persistent-bottom-sheets

**Comment by ahoefling** [comment]
@roubachof thanks for sharing the reference to material design, it is a great article that describes the bottom sheet in detail. Last year I used the current Popup Specification to implement a bottom sheet following the material design. Granted it didn't work that great on iOS because of the rounded corners which is something I hope we (as a community can solve). When I originally implemented the Popup specification in June of 2019, we started dogfooding it internal for all of our popup needs. It is quite a powerful API and EZ did a great job coming up with the original specification that is flexible to solve a lot of problems.

To directly answer some of your questions
> One would expect several states

> it should react to swipe up and down to navigate through state, as well as drag gesture

These are important concerns to get correct. We would need to create a composite control such as a templated view. This template would then support all of the swipe and animation states, while rendering a native popup in the custom renderer which would be referenced via the template. In other words we would create a XCT view that would be a wrapper around the Popup Control.

To be extra clear, I am not saying we don't implement a feature like this, I am suggesting we leverage the power of the Popup Control which will make the implementation details fairly easy. Our focus will be on getting the business rules correct using shared code instead of building a custom renderer.

**Comment by roubachof** [comment]
Thanks for your answer!
Unfortunately I still don't really understand how a native pop up would be relevant for implementing the control we need.
Wouldn't we need just a simple layout with a y translation (this is how google implemented it https://android.googlesource.com/platform/frameworks/support/+/2b53fdcd54f2760ed7957393f9e9f2140105c9ce/design/src/android/support/design/widget/BottomSheetBehavior.java#16)?



Also we need the option to have a dark overlay or not under the bottom sheet, I don't see this option in your spec.

**Comment by IeuanWalker** [comment]
@roubachof would also be good to have a property/ event that can be used to get the %/ or height the sheet is on. Could be used to bind to an animation either on the sheet or on the page behind it.

**Comment by ahoefling** [comment]
@roubachof 
> Wouldn't we need just a simple layout with a y translation

Yes, that would work. I am a big fan of using transparent background in a modal page. That would give us everything we need to do to create the bottom sheet. Based on the information I've seen in this thread, I do think it would be easier to use the native popup as it is just that a layout that isn't attached to the page that you can control the Y translation.

It is really hard to understand if using native popups makes sense or not as there is no specification documented here. I think before we go too far into implementation discussions this needs to be updated into a full XCT specification. At that point I can make my recommendation if a native popup makes sense or another technique. It is okay if don't use a native popup, at this point I still see it being the easiest way to implement this. That opinion may change as more details are documented on features and rules for this.

**Comment by developer9969** [comment]
Hi
I am really glad that this feature request has had very good feedback. This will be a great addition to the toolkit.
I just hope that this process does not drag too long and that is quite a lean process.

@ahoefling I am really blown away that my suggestions seems has had good response, I do not know much about your popup control, I have used the rg.plugin in few projects. Could you look at what I have put together as acceptance criteria.

Having seen the replies in this thread in an ideal world what @roubachof   proposed and the link to the material design would be fantastic. , here it is again https://material.io/components/sheets-bottom/#bottom-sheets-persistent-bottom-sheets   

However I also want to be realistic and if we could deliver something that has potential to grow but acceptable to start with than we could  avoid "not to deliver at all" or takes too long.

Below is my 50 cents:

**MVP (minimum viable product)**
- Bottom sheets should be docked at the bottom
- Bottom sheets should be able to host any content.
- Bottom sheets should work the same in both Android and iOS  
- Bottom sheet should be able to have round corners (pretty standard now)
- Bottom sheets should be independent from the rest of the page/view that is implemented.
- 	If a page has content that makes the screen scrolling down , bottom sheet should alwasy be on top and visible regardless if the main page has content that require scrolling
- 	Some bottom sheet implementation do not work if the screen requires scrolling.	
- Main screen should still be visible when bottom sheet appears
- User should be able to interact with both bottom sheet and main screen
- Tapping on main screen should dismiss the bottom sheet (flag)
- Bottom sheet should be expandable and with a default nudge that can be dragged up and down.
- Bottom sheet can be resized

**Wishlist**

- Bottom sheets can be docked anywhere. e.g top (can come later)
- Theming (can come later)
- In built header with icon - title - description (can come later)


hopefully this gives a starting point for someone to formalize in a spec.

hopes this helps

**Comment by gabsterdev1** [comment]
@ahoefling hi there noticed that you now have added your popup control can you do bottomsheet with your popup? Will have any of the standard bottom sheet feature as highlighted here
[https://material.io/components/sheets-bottom/#bottom-sheets-persistent-bottom-sheets](url)

Thanks

**Comment by ahoefling** [comment]
My company has been dogfooding the popup control for almost 2 years now and we have been using it with a bottom sheet for Android. I know you can implement a bottom sheet for UWP, but there is still a little bit of work that needs to be done for iOS to fully support bottom sheet as the styling has some issues. I am hoping that once it is merged we can identify these and can get some PRs submitted to resolve them. 

If you look at the PR there is a table of supported features between the platform and it documents what is implemented and what couldn't get done.

**Comment by AlleSchonWeg** [comment]
This is also a good component: https://github.com/DIPSAS/DIPS.Xamarin.UI/wiki/Sheet

**Comment by angelru** [comment]
any progress on this?
